### ISSUE TITLE
Integer libraries

### ISSUE THREAD

**Issue by fournet** [FStarLang/FStar#1421]
Is there some other issue for this? I tried to use `FStar.Integers` vs hacl\*'s `Spec.Lib.IntTypes`.

- `Integers` lacks many arithmetic functions, and involve extra annotations. For instance, all the `nat` annotations below are freshly required:
   ```
   module Crypto.HMAC
   open Crypto.Hash 
   open FStar.Integers 

   type bseq = Seq.seq UInt8.t 
   let lbseq (l:nat) = b:bseq {Seq.length b = l}
   let keysized (a:alg13) (l:nat) = l <= maxLength a /\ l + blockLength a < pow2 32

   noextract val hmac: 
     a: alg13 ->
     key: bseq {keysized a (Seq.length key)} ->
     data: bseq {
       let l: nat = Seq.length data in 
       l + blockLength a <= maxLength a } ->
     lbseq (tagLength a)
   ```
  The use of constants is also problematic, e.g. I could not get `2uy + 2uy` to typecheck, with let-binding leading to `bound variable escapes` errors. 

- `IntTypes` circumvent the need for some annotations by using different operators (`+.` instead of `+` etc) for all machine integers. But they are very hacl\*-specific.

**Comment by wintersteiger** [comment]
These work for me:

```
module Scratch

let q =
  let two = 2 in
  two + two

let r =
  let two = 2uy in
  FStar.UInt8.(two +^ two)

open FStar.Integers

let s: int_t Unsigned W8 =
  let two = 2uy in
  two + two
```

The last (`s`) needs the range type annotation, without it I get `Bound variables 'two#85315' escapes; add a type annotation` too.

**Comment by nikswamy** [comment]
Question:

To what extent would people consider heterogeneous overloading of arithmetic operations on machine integers desirable?

E.g., is it important to be able to add a uint32 to a uint16 obtaining a uint32 (with an overflow check, of course).

More generally, @protz  and I have been discussing supporting promotion rules that allow mixing operands of the same signedness but different widths, promoting the operation to work on the largest width. We do not wish to support implicit conversion between types of different signedness.

But, we're not yet sure if this kind of overloading is considered useful in other Everest applications and beyond. On the one hand, it's more flexible; on the other, implicit promotion is less predictable. In any case, it's definitely a departure from the style of the existing FStar.[U]Int[NN] modules.

Feedback most welcome. @fournet @karthikbhargavan @BarryBo @parno  ... ?

**Comment by beurdouche** [comment]
I am not in favor of adding these casts implicitly, the more explicit we are about what we are doing, the better.

**Comment by karthikbhargavan** [comment]
We’ve been working with IntTypes in HACL for a few months now, so here’s a brief experience report.
Of course, our use of integers in HACL is quite specific, so take all of it with a grain of salt.

Casting:
In our code, disallowing implicit casts finds a lot of silly bugs early, so we restrict arithmetic operators to only work on integers of the same type.
More generally, casting from a uint32 to uint64 or the reverse actually requires a machine instruction, and making it explicit is good practice for performant code.
Instead of a generic cast, we find it useful to write type-specific functions like: “to_uint64: #t:inttype -> uint_t t -> uint64_t” (#t gets inferred easily)

Generalizing to Int/Nat:
We tried to extend integers with nat so that we could use +, - etc uniformly for machine integers and for ints/nats.
This does not work well, requiring a bunch of annoying/previously-unnecessary annotations in specs and in value parameters.
More generally, I find it much clearer to have a clean separation between arithmetic in specifications/types vs. code.
So, the discipline we currently follow in HACL* dev branches, and which works quite well is to have 2 categories of integers.

We *only* use int/nat and their subtypes in specs, and we *never* use machine integer operations int specs.
So for example, each spec formula will be of the form “uint_v x + uint_v y == uint_v z”, where x, y, z are machine integers, 
but will never have a formula of the form “x +. y == z”. Similarly, (as far as possible) our value-indexed types are indexed by 
ints/nats and not by machine integers.  

Of course, we also never use int/nat in code, except in ghost computation/implicit parameters, and I know this is not a rule that is followed in non-crypto code.

In summary, based on our experience:
1) I advise against generalizing Integers to include int/nat, and I recommend keeping the spec/code boundary clear by using distinct operators for machine arithmetic.
2) I don’t have a strong opinion about implicit casting, we would probably implement a wrapper that disallows it for our code anyway.

-Karthik

**Comment by nikswamy** [comment]
Thanks Benjamin, Thanks Karthik.

Karthik: About the recommendation against generalizing to integers/nats. I'm actually working actively to try to include a uniform treatment of nats and ints and machine integers. Based on your advice, this may be misguided.

But, I wonder: is your advice more influenced by the overloading not working as well as hoped for, or is it because you really see mathematical and machine integers as separate things with different operators etc. If it is mostly the former, then I hope with a bit of compiler support, we can actually get overloading to work out reasonably well.

Personally, I find having to write and explain the use of `+` vs. `+^` displeasing.

**Comment by karthikbhargavan** [comment]
Hey Nik,
> Karthik: About the recommendation against generalizing to integers/nats. I'm actually working actively to try to include a uniform treatment of nats and ints and machine integers. Based on your advice, this may be misguided.
> 
Like you, I found the idea attractive in principle, so I see why you would be trying to do this.
> But, I wonder: is your advice more influenced by the overloading not working as well as hoped for, or is it because you really see mathematical and machine integers as separate things with different operators etc. If it is mostly the former, then I hope with a bit of compiler support, we can actually get overloading to work out reasonably well.
> 
It is a bit of both. I was originally disappointed with the overloading not working as well as I liked but I think we can probably make this work with some effort, especially if we rule out implicit casting. I would specially experiment with code that uses arithmetic over sequence indices; those are the cases where we were having the most trouble.

After writing and explaining a series of crypto/protocol specs, I find that I do find it easier to think of mathematical integers and concrete machine integers as disjoint. It is useful to point to an operation and syntactically say that this operation is mathematical + and this one is a modular addition. Having said that, if we have a nice core FStar.Integers module which allows us to use + everywhere without annotations, then we can easily wrap around this as needed for various use cases. 

For example, in PQ crypto, we are defining + for integers in Zq, + for polynomials over Zq, + for vectors of polynomials, and + for matrices of polynomials.
It is very useful to overload a single addition operation over all these types, and so I end up defining a new overloaded |+| operator, using roughly the same style as F* integers. 
So far, it appears to work without annotations, but I still keep the vanilla + operator for specs.
> Personally, I find having to write and explain the use of + vs. +^ displeasing
> 
I feel the same way about the notation, especially the ^

-Karthik

**Comment by nikswamy** [comment]
Please see 

https://github.com/FStarLang/FStar/blob/nik_integer_overloading/examples/micro-benchmarks/Test.Integers.fst

Your feedback on the style it promotes and its shortcomings are most welcome.

**Comment by A-Manning** [comment]
I'd be happy to contribute to b. Having implicit resolution controlled by tactics would also mean that we could have user-specified decidable equality, which would improve the currently awkward mechanism for deriving hasEq predicates. Various parts of the standard library could be improved with this kind of mechanism, and we could get more mileage out of lemmas. In general this would be a major improvement to F*, however it would be a fairly large change, and non-trivial to implement.

**Comment by s-zanella** [comment]
From the point of view of users, (3) a. seems best: it's transparent to code not using `FStar.Integers` and doesn't require enabling a flag to get overloading. It also seems the easiest to implement, and allows a seamless transition to (3) b. in the future.

**Comment by parno** [comment]
Historically, we’ve typically written our specifications using int/nat, rather than machine integers, so that the spec isn’t burdened with proof obligations w.r.t. overflow.  In Vale, we experimented with using machine integers (since that seemed natural in a spec about the machine), but there were enough annoying casts, that we swapped to nats and subtypes thereof, which resulted in a cleaner spec.  The proof burden wasn’t too bad, and it’s mainly incurred once for our Hoare-logic layer, so it’s been acceptable thus far.  

In general, I agree with Nik that explaining things like `+^!%` to outsiders is unpleasant, so overloading the standard operators is attractive.  That said, having “surprising” interactions with other libraries (like `len` in the Seq library) seems pretty bad too.  I like Santiago’s proposal, but it would help a lot if F* could produce a more specific error message when the heuristics in 3(a) fail.  Seeing a type unification error when trying to add two integers seems like it will be surprising/confusing to the user, so a helpful error message might go a long way to helping them arrive at a sensible fix.

**Comment by parno** [comment]
@nikswamy As a side note, in the example you linked to, would it be possible to use tactics to automatically push the `ok` check down through the entire computation, rather than having to require `ok` on every sub-term?

**Comment by nikswamy** [comment]
It should be possible to do that using the `synth` meta-program hook. i.e., in a context expecting a `Type` we could write somthing like `synth (gen_ok (quote (z * (x + y)))` instead of `ok (+) x y /\ ok (*) z (x + y)`

**Comment by nikswamy** [comment]
FYI, I implemented 3 a. It works pretty well, though it is still a rather specialized heuristic.

The example is updated and things now compose well between FStar.Integers and Prims.int etc.

https://github.com/FStarLang/FStar/blob/nik_integer_overloading/examples/micro-benchmarks/Test.Integers.fst

**Comment by nikswamy** [comment]
Should be generalize the heuristic to attempt general case splits with backtracking when all else fails. Will implement that soon

**Comment by nikswamy** [comment]
This is now done and the example file is updated: 

https://github.com/FStarLang/FStar/blob/nik_integer_overloading/examples/micro-benchmarks/Test.Integers.fst

I've also removed the ad hoc compiler option (--integer_overloading)

At this point, I think experimenting with FStar.Integers within some of our larger developments would prove some useful battle testing.

Meanwhile, my next step is to add a layer on top of FStar.Integers called FStar.ConstantTime.Integers, providing an abstract type wrapping FStar.Integers and exposing only some operations (the constant time ones). This layer will overload `+, -, ... etc` (all the constant time operations) over both `FStar.Integers.int_t sw` and the abstract type for constant time integers. Of course, operations like `<, <=` etc. will not be provided again at this layer, remaining available only for FStar.Integers.

**Comment by karthikbhargavan** [comment]
> At this point, I think experimenting with FStar.Integers within some of our larger developments would prove some useful battle testing.
> 
Great. I’ll start testing Hacl’s IntTypes via FStar.Integer.
> Meanwhile, my next step is to add a layer on top of FStar.Integers called FStar.ConstantTime.Integers, providing an abstract type wrapping FStar.Integers and exposing only some operations (the constant time ones). This layer will overload +, -, ... etc (all the constant time operations) over both FStar.Integers.int_t sw and the abstract type for constant time integers. Of course, operations like <, <= etc. will not be provided again at this layer, remaining available only for FStar.Integers.
> 
Happy to help with this. Especially for this module, and probably for FStar.Integers as well, it would probably be good to have interfaces to reduce dependencies.

**Comment by prosecco** [comment]
By the way, I don’t fully understand how this works in terms of unification.

When I write (0 + x): F* needs to figure out that 0 has type either int_t Signed Infinite or int_t Unsigned Infinite. 
(a) How does F* decide between Signed/Unsigned?
(b) How does it unify "int_t (s,w)” with “int” in the first place?

Similarly for constants like 0uy, 1ul etc.
Is there some magic I am not understanding here?

**Comment by karthikbhargavan** [comment]
> When I write (0 + x): F* needs to figure out that 0 has type either int_t Signed Infinite or int_t Unsigned Infinite. 
> (a) How does F* decide between Signed/Unsigned?
> (b) How does it unify "int_t (s,w)” with “int” in the first place?

In particular, I would have expected to have to write (0 <: int_t Unsigned Infinite) + x and am wondering how this coercion is inferred.

> 
> Similarly for constants like 0uy, 1ul etc.
> Is there some magic I am not understanding here?
>

**Comment by nikswamy** [comment]
An important part of the redesign of the library is the following design principle:

When defining a type by cases `let t sw = match sw with  | P1 -> t1 ... | Pn -> tn`, things are more robust if each of the `ti` are distinct abstract types.

In particular, things don't work so well if you have something like `let t b = if b then int else nat`. In this situation, if you have `let f #b (x:t b) = ()` then `f 0` is ambiguously typed: should it be `f #true (0 <: int)` or `f #false (0 <: nat)`? This is essentially the question you're asking, I think.

See these parts of the redesign of FStar.Integers:

https://github.com/FStarLang/FStar/blob/master/ulib/FStar.Integers.fst#L51-L67

Reproduced here:
```
type signed_width =
  | Signed of width
  | Unsigned of fixed_width
```

Note, `Unsigned Winfinite` is no longer a valid specifier for the shape of an integer type.

2. Then this bit here, for instance: 
https://github.com/FStarLang/FStar/blob/master/ulib/FStar.Integers.fst#L51-L67

```
let int_t sw : Tot Type0 =
  match sw with
  | Unsigned W8 -> FStar.UInt8.t
  | Unsigned W16 -> FStar.UInt16.t
  | Unsigned W31 -> FStar.UInt31.t
  | Unsigned W32 -> FStar.UInt32.t
  | Unsigned W63 -> FStar.UInt63.t
  | Unsigned W64 -> FStar.UInt64.t
  | Unsigned W128 -> FStar.UInt128.t
  | Signed Winfinite -> int
  | Signed W8 -> FStar.Int8.t
  | Signed W16 -> FStar.Int16.t
  | Signed W31 -> FStar.Int31.t
  | Signed W32 -> FStar.Int32.t
  | Signed W63 -> FStar.Int63.t
  | Signed W64 -> FStar.Int64.t
  | Signed W128 -> FStar.Int128.t
```

See that each case is a distinct abstract type.

3.  Finally:

```
let int       = int_t (Signed Winfinite)

....

let nat = i:int{ 0 <= i }
let pos = i:nat{ 0 < i }
```

With this set up, `0 : Prims.int` as usual.

And `Prims.int` unifies with `int_t ?u` by setting `?u := Signed Winfinite`. This last bit is handled by the new extensions to the unification algorithm, according to the following derivation sketch.

````
   Prims.int ~ int_t ?u
(unfold RHS)
   Prims.int ~ (match ?u with Signed Winfinite -> Prims.int | ... )
(?u := Signed Winfinite, by the unification rule for match)
   Prims.int ~ (match Signed Winfinite with Signed Winfinite -> Prims.int | ... )
(iota reduction)
   Prims.int ~ Prims.int
<==> (refl)
   T
```

**Comment by nikswamy** [comment]
Btw, one further possibility is to overload the machine integer constants themselves, rather than just overloading the operators like `+, -` etc. 

Currently, the integer constants like `0ul` etc. are desugared to `FStar.UInt32.uint_to_t 0 : FStar.UInt32.t`

But, we could instead desugar integer constants `0` to `FStar.Integers.u i : int_t ?w` and let the context determine (via unification) which `0` we really meant. 

This would allow us to write `let f (x:uint32) = x + 1` instead of `let f (x:uint32) = x + 1ul`.

But, I'm not sure this is considered desirable: it has the flavor of an implicit coercion, which, from earlier in the thread, most people responding considered undesirable, at least for coercions among machine integer types. But, maybe for constants this would be nice? I don't know. 

BTW, even without this bit of compiler support, you can already write 

```
let f0 (x:uint_32) = x + u 0
let f1 (x:int_64) = x + u 0
```

With some compiler support, that `u` could be inserted automatically.

**Comment by karthikbhargavan** [comment]
Thanks this was very useful and the test file makes sense to me now.

I particularly like that we can layer these types to extend the overloaded operators.
We’ll happily use some of that in HACL*, especially in specs.

-Karthik

**Comment by nikswamy** [comment]
Hi folks, esp. heads up @karthikbhargavan  @protz  @fournet  @R1kM 

I have a proposal for a simple but general framework for information-flow control and constant-time integers. 

I am trying to address a long-standing requirement that we support more than just one kind of secret, e.g., to distinguish between secret plain texts and secret keys; or to distinguish between plain texts that belong to different principals.

Please see
1. https://github.com/FStarLang/FStar/blob/nik_integer_overloading/ulib/FStar.IFC.fst
2. https://github.com/FStarLang/FStar/blob/nik_integer_overloading/ulib/FStar.ConstantTime.Integers.fst
(and 3. https://github.com/FStarLang/FStar/blob/nik_integer_overloading/examples/micro-benchmarks/Test.IFC.fst)

In (1), I define a basic abstraction in FStar.IFC for information flow control on pure programs

In (2), I use FStar.IFC and define `FStar.ConstantTime.Integers.secret_int l s`, the type of a secret machine integer (of signedness/width described by `s`) labeled at level `l` (a lattice element).


Note, I am focusing mainly on confidentiality here, not integrity.

If you buy this general style, I will complete FStar.ConstantTime.Integers to provide more constant-time operations (currently only `+` and `+%`, though the others should be similar).

I then plan to add another overloading layer to multiplex operations between FStar.ConstantTime.Integers and FStar.Integers. 

Thanks for taking a look ... comments welcome!

**Comment by nikswamy** [comment]
Added an interface with a multiplexing layer and some examples at the end of this file:

https://github.com/FStarLang/FStar/blob/nik_integer_overloading/ulib/FStar.ConstantTime.Integers.fsti

**Comment by karthikbhargavan** [comment]
I like this style, especially for coding up easy-to-understand high-level secure APIs (IFC will be more understandable, in my opinion, than complex indexed types.)

For low-level APIs like integers, however, I am not fully sold yet.

Instead of "hard-coding" a lattice, I wonder if it would not be better to have indexed integers at an arbitrary index, without promotion:

val secret_int (#idx_t:Type0)
               (l:idx_t)
               (s:sw) : Type0

All operations would preserve indices:

val addition (#idx_t:Type0)
             (#l:idx_t)
             (#s:sw)
             (x : secret_int l s)
             (y : secret_int l s {ok ( + ) (m x) (m y)})
    : Tot (z:secret_int l s{m z == m x + m y})

And, by default, there is no promotion/declassification between indices, except you can coerce  raw F* integers into secret_integers:

val u:    (#idx_t:Type0)    
             (#l:idx_t)
             (#s:sw)
             (i:int_t s)
       : Tot (z:secret_int l s{m z == Integers.v i})

For crypto-like examples, I think this is enough, and for miTLS I think the arbitrary index is more useful than defining a lattice, but maybe I am missing something.

**Comment by nikswamy** [comment]
Hi Karthik, 

Thanks for the comments.

A few responses: 

1. As a general design principle, following the recommendation of comments such as this (https://github.com/FStarLang/FStar/issues/1421#issuecomment-392872316), the interface provided by all the integer modules is homogeneous in the indexes, i.e., no implicit casts/promotions. This is also true of ConstantTime.Integers. See, e.g.,  https://github.com/FStarLang/FStar/blob/nik_integer_overloading/ulib/FStar.ConstantTime.Integers.fsti#L77

As such, for typical usage, the interface already has the "preserve indices" shape that you prefer.

A `promote` operations for lifting the IFC index is available explicitly, but one needn't use it.

2. One requirement for using this library is that your indices must be arranged in a semilattice. But, I hope this requirement is not too hard to satisfy, so long as your index type supports decidable equality. For instance, if you want to index using a type `t`, you can make a trivial semi-lattice out of it by indexing using `option t` instead and defining 

```
SemiLattice #(option t) None (fun x y -> if x=y then x else None)
```

Would you really want to index by types that do not support equality?

If so, I could easily also adapt the library defining some sort of partial semi_lattice, so that the join and the top element are not always required to exists, but that if they do, it satisfies the laws you'd expect.

3. Label polymorphic code should just work: I've now set things up so that the lattice and the labels etc. get erased. See https://github.com/FStarLang/FStar/blob/nik_integer_overloading/examples/micro-benchmarks/Test.ConstantTime.Integers.fst

In particular, label polymorphic code would look like this:

```
let ex_poly #sl (#l:lattice_element sl) (x:l_uint32 l) (y:l_uint32 l) = x +% y
```

It's pretty close in shape to the kind of type you write (same implicit argument arity etc.), except instead of indexing by any type, I am indexing by a type drawn from a semi lattice.

Extracts in OCaml to

```

//NS: label-polymorphic code is extracted just like
//    the monomorphic `ex1`,
//    except we have additional unit arguments for the erased
//    lattice and lattice element.
//    Kremlin should be able to remove these redundant arguments, though.
let (ex_poly :
  unit ->
    unit ->
      (unit,unit,unit) FStar_ConstantTime_Integers.secret_int ->
        (unit,unit,unit) FStar_ConstantTime_Integers.secret_int ->
          (unit,unit,unit) FStar_ConstantTime_Integers.secret_int)
  =
  fun sl  ->
    fun l  ->
      fun x  ->
        fun y  ->
          FStar_ConstantTime_Integers.addition_mod () ()
            (FStar_Integers.Unsigned FStar_Integers.W32) x y
```

4. While IFC label polymorphic code shoould just work, extracting code that is polymorphic on the signedness & width of a machine integer will not really work in KreMLin, since this kind of dependent type is not expressible in C (we'd have to generated a tagged union on the fly and that's not something we support). In any case, I doubt you'd want to write efficient low-level code in this style. That said, you're free to write this kind of integer-generic code in specs.
